networkx/drawing/layout.py

```diff
@@ -617,3 +619,178 @@ def flatten(l):
             yield el
 
+
+def kamada_kawai(G, pos=None, K=1.0, L=0.1, iterations=None,
+                 tolerance=1, D=0.0, weight='weight'):
+    """Position nodes using Kamada-Kawai force-directed algorithm.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    pos : dict or None  optional (default=None)
+       Initial positions for nodes as a dictionary with node as keys
+       and values as a list or tuple.  If None, then use random initial
+       positions.
+
+    K : float (default=1.0)
+       A constant which determines the strength of the spring.
+
+    L : float (default=0.1)
+        L a constant which determines desirable
+        length of a single edge in the display plane.
+
+    iterations : int  optional (default=None)
+       Number of maximum iterations of the algorithm.
+       If None, then 10 * n iterations will be completed.
+
+    tolerance : int  optional (default=1)
+       Number of maximum iterations of the inner loop of the
+       algorithm.
+
+    D : float  optional (default=0.0)
+       A constant which determines the minimum desirable value of
+       energy. If energy's value is lower of this constant then
+       algorithm terminates.
+
+    weight : string or None   optional (default='weight')
+        The edge attribute that holds the numerical value used for
+        the edge weight.  If None, then all edge weights are 1.
+
+    Returns
+    -------
+    dict :
+       A dictionary of positions keyed by node
+
+    Examples
+    --------
+    >>> G=nx.path_graph(4)
+    >>> pos=nx.kamada_kawai(G)
+    """
+    try:
+        import numpy as np
+    except ImportError:
+        raise ImportError("kamada_kawai() requires numpy: http://scipy.org/ ")
+
+    n = G.number_of_nodes()
+    number_of_derivatives = 5
+    if n == 0:
+        return {}
+    if n == 1:
+        return {G.nodes()[0]: [0, 0]}
+
+    dist = nx.floyd_warshall_numpy(G, weight=weight)
+    diameter = np.ma.masked_invalid(dist).max()
+
+    # Initialize parameters
+    if iterations is None:
+        iterations = 10 * n
+    if tolerance is None:
+        tolerance = 1
+
+    if pos is None:
+        pos_arr = np.asarray(np.random.random((n, 2)), dtype=dist.dtype)
+    else:
+        # Determine size of existing domain to adjust initial positions
+        dom_size = max(flatten(pos.values()))
+        shape = (n, 2)
+        pos_arr = np.random.random(shape) * dom_size
+        for i, u in enumerate(G):
+            if u in pos_arr:
+                pos_arr[i] = np.asarray(pos[u])
+
+    # Initialize k and l arrays
+    k = np.zeros(dist.shape, dtype=dist.dtype)
+    l = np.zeros(dist.shape, dtype=dist.dtype)
+    for u in range(n):
+        for v in range(n):
+            if u != v:
+                if math.isinf(dist[u, v]):
+                    dist[u, v] = diameter
+                k[u, v] = K / (dist[u, v] ** 2.0)
+                l[u, v] = L * dist[u, v]
+
+    def compute_derivatives(pxx, pxy, pyx, pyy, kappa, la):
```

Comment:
In any case, this function should probably be defined at module-level, not within another function. "Flat is better than nested."